clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -369,6 +434,94 @@ public void logUnused() {
         return objects;
     }
 
+    private Map<String, String> getMapAsString(Map<?, ?>  configMap) {
+        Map<String, String> configMapAsString = new HashMap<>();
+        for (Map.Entry<?, ?> entry : configMap.entrySet()) {
+            if (!(entry.getKey() instanceof String))
+                throw new ConfigException(entry.getKey().toString(), entry.getValue(),
+                    "Key must be a string.");
+            if (entry.getValue() instanceof String)
+                configMapAsString.put((String) entry.getKey(), (String) entry.getValue());
```

Comment:
Is the intention to skip values that are not strings because they can't be variables? Might be worth a comment here if that's the case.

I also wonder if it'd be better to keep the original code that is checking that strings are keys in the constructor, and use this method to separately loop through the entries. This doesn't happen that often, so the extra loop is not that big a deal, but this helps reduce the risk of changing behavior.

Essentially, the constructor would start with:
```
        for (Map.Entry<?, ?> entry : originals.entrySet())
            if (!(entry.getKey() instanceof String))	
                 throw new ConfigException(entry.getKey().toString(), entry.getValue(), "Key must be a string.");	                

        this.originals = resolveConfigVariables(originals, configProviderProps, (Map<String, ?>) originals);
```

Note that because the call to `getMapAsString` method only returns the subset of the originals that might have variables, it's probably better to call that within the `resolveConfigVariables(...)` method to encapsulate the logic a bit more.

Then this method could be:
```
   private Map<String, String> getMapAsString(Map<?, ?>  configMap) {
        Map<String, String> configMapAsString = new HashMap<>();
        for (Map.Entry<?, ?> entry : configMap.entrySet()) {
            if (entry.getValue() instanceof String) {
                // We only need the values that are strings, because only they can be variables
                configMapAsString.put((String) entry.getKey(), (String) entry.getValue());
            }
        }
        return configMapAsString;
   }
```

Reply:
Yes as variables can only be strings .. I have added the comment and fix